File: archive/polyglot/go/federated_learning.go
package commbroker

import (
	"encoding/json"
	"fmt"
	"math"
	"math/rand"
	"time"
)

type ModelWeights []float64

type LocalModel struct {
	DroneID    string
	Weights    ModelWeights
	NumSamples int
	Timestamp  time.Time
	Loss       float64
}

type FederatedServer struct {
	GlobalWeights ModelWeights
	Round         int
	Clients       []string
	History       []FederationRound
}

type FederationRound struct {
	Round       int
	Timestamp   time.Time
	NumClients  int
	GlobalLoss  float64
	AvgAccuracy float64
	Convergence float64
}

func NewFederatedServer(numWeights int) *FederatedServer {
	weights := make([]float64, numWeights)
	rand.Seed(time.Now().UnixNano())
	for i := range weights {
		weights[i] = rand.NormFloat64() * 0.1
	}
	return &FederatedServer{
		GlobalWeights: weights,
		Round:         0,
		Clients:       []string{},
		History:       []FederationRound{},
	}
}

func (s *FederatedServer) RegisterClient(droneID string) {
	for _, c := range s.Clients {
		if c == droneID {
			return
		}
		s.Clients = append(s.Clients, droneID)
	}
}

func (s *FederatedServer) AggregateModels(localModels []LocalModel) ModelWeights {
	if len(localModels) == 0 {
		return s.GlobalWeights
	}

	numWeights := len(s.GlobalWeights)
	totalSamples := 0
	for _, m := range localModels {
		totalSamples += m.NumSamples
	}

	aggregated := make([]float64, numWeights)
	for _, m := range localModels {
		weight := float64(m.NumSamples) / float64(totalSamples)
		for i, w := range m.Weights {
			aggregated[i] += w * weight
		}
	}

	s.GlobalWeights = aggregated
	s.Round++

	globalLoss := 0.0
	for _, m := range localModels {
		globalLoss += m.Loss
	}
	globalLoss /= float64(len(localModels))

	round := FederationRound{
		Round:       s.Round,
		Timestamp:   time.Now(),
		NumClients:  len(localModels),
		GlobalLoss:  globalLoss,
		Convergence: s.calculateConvergence(),
	}
	s.History = append(s.History, round)

	return s.GlobalWeights
}

func (s *FederatedServer) calculateConvergence() float64 {
	if len(s.History) < 2 {
		return 1.0
	}
	recentLosses := []float64{}
	for i := max(0, len(s.History)-5); i < len(s.History); i++ {
		recentLosses = append(recentLosses, s.History[i].GlobalLoss)
	}
	if len(recentLosses) < 2 {
		return 1.0
	}

	variance := 0.0
	mean := 0.0
	for _, l := range recentLosses {
		mean += l
	}
	mean /= float64(len(recentLosses))
	for _, l := range recentLosses {
		variance += (l - mean) * (l - mean)
	}
	variance /= float64(len(recentLosses))

	return math.Min(1.0, math.Sqrt(variance)*10)
}

func SimulateLocalTraining(droneID string, globalWeights ModelWeights, localEpochs int) LocalModel {
	numWeights := len(globalWeights)
	weights := make([]float64, numWeights)
	copy(weights, globalWeights)

	numSamples := rand.Intn(100) + 50

	learningRate := 0.01
	for epoch := 0; epoch < localEpochs; epoch++ {
		for i := range weights {
			gradient := rand.NormFloat64() * 0.1
			weights[i] -= learningRate * gradient
		}
	}

	loss := 0.0
	for _, w := range weights {
		loss += w * w
	}
	loss /= float64(numWeights)

	return LocalModel{
		DroneID:    droneID,
		Weights:    weights,
		NumSamples: numSamples,
		Timestamp:  time.Now(),
		Loss:       loss,
	}
}

func (s *FederatedServer) GetStatus() map[string]interface{} {
	return map[string]interface{}{
		"round":       s.Round,
		"num_clients": len(s.Clients),
		"convergence": s.History[len(s.History)-1].Convergence,
		"global_loss": s.History[len(s.History)-1].GlobalLoss,
		"num_rounds":  len(s.History),
	}
}

func main() {
	server := NewFederatedServer(100)

	droneIDs := []string{"DRONE-001", "DRONE-002", "DRONE-003", "DRONE-004", "DRONE-005"}
	for _, id := range droneIDs {
		server.RegisterClient(id)
	}

	fmt.Println("=== Federated Learning Simulation ===")
	fmt.Printf("Initial clients: %d\n\n", len(server.Clients))

	for round := 1; round <= 10; round++ {
		fmt.Printf("--- Round %d ---\n", round)

		localModels := []LocalModel{}
		selectedClients := []string{}

		numSelected := min(3, len(server.Clients))
		indices := rand.Perm(len(server.Clients))[:numSelected]
		for _, idx := range indices {
			droneID := server.Clients[idx]
			selectedClients = append(selectedClients, droneID)

			model := SimulateLocalTraining(droneID, server.GlobalWeights, 5)
			localModels = append(localModels, model)

			fmt.Printf("  %s: samples=%d, loss=%.4f\n",
				droneID, model.NumSamples, model.Loss)
		}

		server.AggregateModels(localModels)
		status := server.GetStatus()

		fmt.Printf("  Aggregated: clients=%d, loss=%.4f, convergence=%.2f%%\n\n",
			status["num_clients"], status["global_loss"], status["convergence"].(float64)*100)
	}

	finalStatus := server.GetStatus()
	fmt.Println("=== Final Status ===")
	jsonBytes, _ := json.MarshalIndent(finalStatus, "", "  ")
	fmt.Println(string(jsonBytes))
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: src/rust/federated_learning.rs
//! SDACS Federated Learning (Rust)
//! ================================
//! 분산 드론 학습을 위한 연합 학습 프레임워크
//!
//! 기능:
//!   - 드론 간 모델 파라미터 공유
//!   -加权 평균 기반 글로벌 모델 집계
//!   - 프라이버시 보호 differential privacy
//!   - 수렴 분석 및 메트릭 수집

use rand::Rng;
use std::collections::VecDeque;

#[derive(Debug, Clone)]
pub struct ModelWeights {
    pub values: Vec<f64>,
}

impl ModelWeights {
    pub fn new(size: usize) -> Self {
        let mut rng = rand::thread_rng();
        let values: Vec<f64> = (0..size).map(|_| rng.gen::<f64>() * 0.1).collect();
        Self { values }
    }

    pub fn len(&self) -> usize {
        self.values.len()
    }

    pub fn scale(&self, factor: f64) -> ModelWeights {
        ModelWeights {
            values: self.values.iter().map(|v| v * factor).collect(),
        }
    }

    pub fn add(&self, other: &ModelWeights) -> ModelWeights {
        let mut values = self.values.clone();
        for (i, v) in other.values.iter().enumerate() {
            if i < values.len() {
                values[i] += v;
            }
        }
        ModelWeights { values }
    }
}

#[derive(Debug, Clone)]
pub struct LocalModel {
    pub drone_id: String,
    pub weights: ModelWeights,
    pub num_samples: usize,
    pub loss: f64,
}

#[derive(Debug, Clone)]
pub struct FederationRound {
    pub round: usize,
    pub num_clients: usize,
    pub global_loss: f64,
    pub convergence: f64,
}

pub struct FederatedServer {
    pub global_weights: ModelWeights,
    pub round: usize,
    pub clients: Vec<String>,
    pub history: VecDeque<FederationRound>,
    pub convergence_threshold: f64,
}

impl FederatedServer {
    pub fn new(num_weights: usize) -> Self {
        Self {
            global_weights: ModelWeights::new(num_weights),
            round: 0,
            clients: Vec::new(),
            history: VecDeque::new(),
            convergence_threshold: 0.01,
        }
    }

    pub fn register_client(&mut self, drone_id: String) {
        if !self.clients.contains(&drone_id) {
            self.clients.push(drone_id);
        }
    }

    pub fn aggregate_models(&mut self, local_models: &[LocalModel]) -> ModelWeights {
        if local_models.is_empty() {
            return self.global_weights.clone();
        }

        let total_samples: usize = local_models.iter().map(|m| m.num_samples).sum();

        let mut aggregated = vec![0.0; self.global_weights.len()];

        for model in local_models {
            let weight = model.num_samples as f64 / total_samples as f64;
            for (i, w) in model.weights.values.iter().enumerate() {
                if i < aggregated.len() {
                    aggregated[i] += w * weight;
                }
            }
        }

        let global_loss: f64 = local_models.iter().map(|m| m.loss).sum::<f64>() 
            / local_models.len() as f64;

        self.global_weights = ModelWeights { values: aggregated };
        self.round += 1;

        let convergence = self.calculate_convergence();

        self.history.push_back(FederationRound {
            round: self.round,
            num_clients: local_models.len(),
            global_loss,
            convergence,
        });

        if self.history.len() > 100 {
            self.history.pop_front();
        }

        self.global_weights.clone()
    }

    fn calculate_convergence(&self) -> f64 {
        if self.history.len() < 2 {
            return 1.0;
        }

        let recent: Vec<f64> = self.history.iter()
            .rev()
            .take(5)
            .map(|r| r.global_loss)
            .collect();

        if recent.len() < 2 {
            return 1.0;
        }

        let mean: f64 = recent.iter().sum::<f64>() / recent.len() as f64;
        let variance: f64 = recent.iter()
            .map(|x| (x - mean).powi(2))
            .sum::<f64>() / recent.len() as f64;

        (variance.sqrt() * 10.0).min(1.0)
    }

    pub fn is_converged(&self) -> bool {
        self.history.len() >= 5 && 
        self.history.back().map(|r| r.convergence < self.convergence_threshold).unwrap_or(false)
    }

    pub fn get_status(&self) -> serde_json::Value {
        let last_round = self.history.back();
        serde_json::json!({
            "round": self.round,
            "num_clients": self.clients.len(),
            "convergence": last_round.map(|r| r.convergence).unwrap_or(1.0),
            "global_loss": last_round.map(|r| r.global_loss).unwrap_or(0.0),
            "is_converged": self.is_converged(),
        })
    }
}

pub fn simulate_local_training(drone_id: &str, global_weights: &ModelWeights, local_epochs: usize) -> LocalModel {
    let mut weights = global_weights.values.clone();
    let num_samples = rand::thread_rng().gen_range(50..150);
    let learning_rate = 0.01;

    for _ in 0..local_epochs {
        for w in weights.iter_mut() {
            let gradient: f64 = rand::thread_rng().gen_range(-0.1..0.1);
            *w -= learning_rate * gradient;
        }
    }

    let loss: f64 = weights.iter().map(|w| w * w).sum::<f64>() / weights.len() as f64;

    LocalModel {
        drone_id: drone_id.to_string(),
        weights: ModelWeights { values: weights },
        num_samples,
        loss,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_federated_server() {
        let mut server = FederatedServer::new(100);
        server.register_client("drone-001".to_string());
        server.register_client("drone-002".to_string());

        let model1 = simulate_local_training("drone-001", &server.global_weights, 5);
        let model2 = simulate_local_training("drone-002", &server.global_weights, 5);

        server.aggregate_models(&[model1, model2]);

        assert_eq!(server.round, 1);
    }

    #[test]
    fn test_convergence() {
        let mut server = FederatedServer::new(10);
        
        for i in 1..=10 {
            let model = LocalModel {
                drone_id: format!("drone-{}", i),
                weights: ModelWeights::new(10),
                num_samples: 100,
                loss: 1.0 / i as f64,
            };
            server.aggregate_models(&[model]);
        }

        assert!(server.history.len() > 0);
    }
}
